README.md
# Lockit utilities

[![Build Status](https://travis-ci.org/zemirco/lockit-utilities.svg?branch=master)](https://travis-ci.org/zemirco/lockit-utilities) [![NPM version](https://badge.fury.io/js/lockit-utils.svg)](http://badge.fury.io/js/lockit-utils)

Small utilities module for [lockit](https://github.com/zemirco/lockit).

## Installation

`npm install lockit-utils`

```js
var utls = require('lockit-utils');
```

## Configuration

```js
// redirect target when requesting restricted page
exports.login = {
  route: '/login'
};

// database connection string
// CouchDB
exports.db = 'http://127.0.0.1:5984/';

// MongoDB
// exports.db = {
//   url: 'mongodb://127.0.0.1/',
//   name: 'test',
//   collection: 'users'
// };

// PostgreSQL
// exports.db = {
//   url: 'postgres://127.0.0.1:5432/',
//   name: 'users',
//   collection: 'my_user_table'
// };

// MySQL
// exports.db = {
//   url: 'mysql://127.0.0.1:3306/',
//   name: 'users',
//   collection: 'my_user_table'
// };

// SQLite
// exports.db = {
//   url: 'sqlite://',
//   name: ':memory:',
//   collection: 'my_user_table'
// };
```

## Features

- protect routes from unauthorized access and redirect
- get database and lockit adapter from connection string
- generate link to QR code image for two-factor auth
- verify provided two-factor token
- destroy a session (works with cookie sessions and session stores)

## Methods
### restrict([config])

Prevent users who aren't logged-in from accessing routes.
Use `login.route` for redirection. Function also remembers the requested url
and user is redirected after successful login. If `rest` is enabled
you'll get a `401` response.


- `config` **Object** *optional*  - Configuration object

  - `login` **String** - Route that handles the login process - default `'/login'`





#### Example


`config.js`
```javascript
exports.login = {
  route: '/login'
};
```

`app.js`
```javascript
var config = require('./config.js');
app.get('/private', utils.restrict(config), function(req, res) {
  res.send('only a logged in user can see this');
})
```



### getDatabase(config)

Get type of database and database adapter name from connection information.


- `config` **Object** - Configuration object

  - `db` **String, Object** - Database connection string / object



#### Returns

- **Object** - Object containing database `type` and `adapter`



#### Example


`config.js (CouchDB)`
```javascript
exports.db = 'http://127.0.0.1:5984/';
```

`config.js (all other DBs)`
```javascript
exports.db = {
  url: 'postgres://127.0.0.1:5432/',
  name: 'users',
  collection: 'my_user_table'
}
```

`app.js`
```javascript
var config = require('./config.js');
var db = util.getDatabase(config);
// {
//   type: 'couchdb',
//   adapter: 'lockit-couchdb-adapter'
// }
```



### qr(config)

Generate link to QR code,  uses <a href="https://developers.google.com/chart/infographics/docs/qr_codes">Google Charts</a>.


- `config` **Object** - Configuration object

  - `key` **String** - Individual random key for user

  - `email` **String** - User email for Google Authenticator app

  - `issuer` **String** - Issuer for Google Authenticator - default `'Lockit'`



#### Returns

- **String** - URL for QR code



#### Example


```javascript
var config = {
  key: 'abcd1234',
  email: 'mirco.zeiss@gmail.com'
};
var link = util.qr(config);
// https://chart.googleapis.com/chart?chs=200x200&cht=qr&chl=otpauth%3A%2F%2Ftotp%2FLockit%3Amirco.zeiss%40gmail.com%3Fsecret%3DMFRGGZBRGI2DI%3D%3D%3D%26issuer%3DLockit
```



### verify(token, key, [options])

Verify a two-factor authentication token, uses <a href="http://en.wikipedia.org/wiki/Time-based_One-time_Password_Algorithm">time-based one-time password algorithm (totp)</a>.
To be used with <a href="https://support.google.com/accounts/answer/1066447?hl=en">Google Authenticator</a>.


- `token` **String** - The two-factor token to verify

- `key` **String** - The individual key for the user

- `options` **Object** *optional*  - Options object for
  <a href="https://github.com/guyht/notp#totpverifytoken-key-opt">notp#totp.verify</a>

  - `window` **String** - Allowable margin for counter - default `6`

  - `time` **Number** - Time step of counter in seconds - default `30`



#### Returns

- **Boolean** - `true` if token is valid



#### Example


```javascript
var key = 'abcd1234';
var token = '236709';
var valid = util.verify(token, key);
if (valid) {
  // continue here
}
```



### destroy(req, done)

Destroy the current session. Works with cookie sessions and session stores.


- `req` **Object** - The default Express request object

- `done` **function** - Function executed when session is destroyed





#### Example


```javascript
util.destroy(req, function() {
  // user is now logged out
});
```



### pipe(source, target)

Pipe events from `source` to `target`. `source` can be a single event
emitter or an Array of event emitters.


- `source` **Object, Array** - Single event emitter or Array of event emitters

- `target` **Object** - Single event emitter





#### Example


```javascript
var util = require('util');
var events = require('events');
var utils = require('lockit-utils');

var Child = function() {};
util.inherits(Child, events.EventEmitter);

var Mother = function() {};
util.inherits(Mother, events.EventEmitter);

var child = new Child();
var mother = new Mother();

utils.pipe(child, mother);

mother.on('action', function(action) {
  console.log('look the child is ' + action);
});

child.emit('action', 'smiling');
```


## Test

`make test`

## License

MIT


History.md

##### 0.5.3 / 2015-07-03

- fix test command in `package.json`

##### 0.5.2 / 2015-07-03

- use eslint
- update dependencies
- use make instead grunt

##### 0.5.1 / 2014-08-18

- use `req.originalUrl` instead `req.url` for redirect
- update dependencies

##### 0.5.0 / 2014-05-28

- add `pipe` method to pipe events from child modules to parent module

##### 0.4.2 / 2014-05-23

- use `req.session.regenerate` instead `req.session.destroy` (fix [#1](https://github.com/zemirco/lockit-utilities/issues/1))

##### 0.4.1 / 2014-05-23

- fix `restrict` by using `config.login.route` instead `config.loginRoute`
- drop unnecessary development dependencies
- update docs

##### 0.4.0 / 2014-05-05

- improve documentation
- `restrict()` function changed

  Instead of checking for `req.session.name` and `req.session.email`
  it now checks for `req.session.loggedIn`.

- add function `qr()` to generate link to QR code image

  ```js
  var config = {
    key: 'abcd1234',
    email: 'mirco.zeiss@gmail.com'
  };
  var link = util.qr(config);
  // https://chart.googleapis.com/chart?chs=200x200&cht=qr&chl=otpauth%3A%2F%2Ftotp%2FLockit%3Amirco.zeiss%40gmail.com%3Fsecret%3DMFRGGZBRGI2DI%3D%3D%3D%26issuer%3DLockit
  ```

- add function `verify()` to verify two-factor authentication token

  ```js
  var key = 'abcd1234';
  var token = '236709';
  var valid = util.verify(token, key);
  if (valid) {
    // continue here
  }
  ```

- add function `destroy()` to destroy sessions

  ```js
  util.destroy(req, function() {
    // user is now logged out
  });
  ```

- split tests into multiple files
- update development dependencies

##### 0.3.0 / 2014-04-17

- make `getDatabase()` work with String and Object


test/qr.js
'use strict';

var should = require('should');
var utls = require('../index.js');

describe('qr', function() {

  it('should return a String', function() {
    var config = {
      key: 'abcd1244',
      email: 'mirco.zeiss@gmail.com'
    };
    var link = utls.qr(config);
    link.should.be.type('string');
  });

  it('should use "Lockit" as default issuer', function() {
    var config = {
      key: 'abcd1244',
      email: 'mirco.zeiss@gmail.com'
    };
    var link = utls.qr(config);
    var label = encodeURIComponent('Lockit:mirco.zeiss@gmail.com');
    var issuer = encodeURIComponent('&issuer=Lockit');
    link.should.containEql(label);
    link.should.containEql(issuer);
  });

  it('should allow for a custom issuer', function() {
    var config = {
      key: 'abcd1244',
      email: 'mirco.zeiss@gmail.com',
      issuer: 'app'
    };
    var link = utls.qr(config);
    var label = encodeURIComponent('app:mirco.zeiss@gmail.com');
    var issuer = encodeURIComponent('&issuer=app');
    link.should.containEql(label);
    link.should.containEql(issuer);
  });

  it('should use base32 encoding for the key', function() {
    var config = {
      key: 'abcd1244',
      email: 'mirco.zeiss@gmail.com'
    };
    var encoded = encodeURIComponent('MFRGGZBRGI2DI===');
    var link = utls.qr(config);
    link.should.containEql(encoded);
  });

});


test/pipe.js
'use strict';

var util = require('util');
var events = require('events');
var should = require('should');
var utils = require('../index.js');

describe('pipe', function() {

  it('should work with single event emitter', function(done) {
    var Mother = function(){};
    util.inherits(Mother, events.EventEmitter);
    var Child = function(){};
    util.inherits(Child, events.EventEmitter);
    var mother = new Mother();
    var child = new Child();
    utils.pipe(child, mother);
    mother.on('smile', function(one, two, three) {
      one.should.equal('a');
      two.should.equal('b');
      three.should.equal('c');
      done();
    });
    child.emit('smile', 'a', 'b', 'c');
  });

  it('should work with an Array of event emitters', function(done) {
    var Mother = function(){};
    util.inherits(Mother, events.EventEmitter);
    var Child = function(){};
    util.inherits(Child, events.EventEmitter);
    var Teen = function(){};
    util.inherits(Teen, events.EventEmitter);
    var mother = new Mother();
    var child = new Child();
    var teen = new Teen();
    utils.pipe([child, teen], mother);
    var called = 0;
    mother.on('smile', function() {
      called++;
      if (called === 2) {done(); }
    });
    child.emit('smile');
    teen.emit('smile');
  });

});


test/test.js

var should = require('should');
var utls = require('../index.js');

describe('lockit-utils', function() {

  describe('restrict', function() {

    it('should protect routes from unauthorized access', function(done) {
      var req = {
        session: {
          name: null,
          email: null
        }
      };
      var fn = utls.restrict();
      var res = {
        redirect: function(route) {
          // simply make sure res.redirect() is called
          done();
        }
      };
      fn(req, res, function() {});
    });

    it('should allow access to authorized users', function(done) {
      var req = {
        session: {
          name: 'john',
          email: 'john@email.com'
        }
      };
      var res = {};
      var fn = utls.restrict();
      // simply make sure next() is called
      fn(req, res, function() {
        done();
      });
    });

    it('should use the default route when none is specified', function(done) {
      var req = {
        session: {
          name: null,
          email: null
        },
        url: '/profile'
      };
      var fn = utls.restrict();
      var res = {
        redirect: function(route) {
          route.should.equal('/login?redirect=/profile');
          done();
        }
      };
      fn(req, res, function() {});
    });

    it('should use the custom route if one is specified', function(done) {
      var req = {
        session: {
          name: null,
          email: null
        },
        url: '/profile'
      };
      var config = {
        loginRoute: '/test'
      };
      var fn = utls.restrict(config);
      var res = {
        redirect: function(route) {
          route.should.equal('/test?redirect=/profile');
          done();
        }
      };
      fn(req, res, function() {});
    });

    it('should send 401 if rest is active', function(done) {
      var req = {
        session: {
          name: null,
          email: null
        }
      };
      var config = {
        rest: true
      };
      var fn = utls.restrict(config);
      var res = {
        send: function(status) {
          status.should.equal(401);
          done();
        }
      };
      fn(req, res, function() {});
    });

  });

  describe('getDatabaseType', function() {

    it('should work for CouchDB', function(done) {
      var config = {
        db: 'http://127.0.0.1:5984/test'
      };
      var db = utls.getDatabase(config);
      db.type.should.equal('couchdb');
      db.adapter.should.equal('lockit-couchdb-adapter');
      var config_alt = {
        db: {
          url: 'http://127.0.0.1:5984/test'
        }
      };
      var db_alt = utls.getDatabase(config_alt);
      db_alt.type.should.equal('couchdb');
      db_alt.adapter.should.equal('lockit-couchdb-adapter');
      done();
    });

    it('should work for MongoDB', function(done) {
      var config = {
        db: {
          url: 'mongodb://127.0.0.1/test'
        }
      };
      var db = utls.getDatabase(config);
      db.type.should.equal('mongodb');
      db.adapter.should.equal('lockit-mongodb-adapter');
      done();
    });

    it('should work for PostgreSQL', function(done) {
      var config = {
        db: {
          url: 'postgres://127.0.0.1:5432/users'
        }
      };
      var db = utls.getDatabase(config);
      db.type.should.equal('postgresql');
      db.adapter.should.equal('lockit-sql-adapter');
      done();
    });

    it('should work for MySQL', function(done) {
      var config = {
        db: {
          url: 'mysql://127.0.0.1:9821/users'
        }
      };
      var db = utls.getDatabase(config);
      db.type.should.equal('mysql');
      db.adapter.should.equal('lockit-sql-adapter');
      done();
    });

    it('should work for SQLite', function(done) {
      var config = {
        db: {
          url: 'sqlite://'
        }
      };
      var db = utls.getDatabase(config);
      db.type.should.equal('sqlite');
      db.adapter.should.equal('lockit-sql-adapter');
      done();
    });

  });

});


generate.js

var fs = require('fs');
var handlebars = require('handlebars');
var docs = require('./docs.json');

/**
 * Internal helper functions
 */

String.prototype.p = function() {
  return this.replace(/<\/?p>/g, '');
};

String.prototype.code = function() {
  return this.replace(/<\/?code>/g, '`');
};

/**
 * Handlebars helpers
 */

// parameters inside function(...)
handlebars.registerHelper('param', function(param, first) {
  // ignore params like config.key
  if (param.name.indexOf('.') !== -1) return;
  // check if param is optional
  var res = param.optional ? '[' + param.name + ']' : param.name;
  // no comma before first parameter
  if (!first) res = ', ' + res;
  return res;
});

// generate example code blocks with captions
handlebars.registerHelper('example', function(example) {
  // get caption
  var captionRe = /<caption>(.+)<\/caption>/;
  var match = captionRe.exec(example);
  var caption = (match) ? '`' + match[1] + '`\n' : '';
  // remove caption from code
  var codeRe = /<caption>(.+)<\/caption>\r\n\s{2}/;
  var code = example.replace(codeRe, '');
  // fix bug caused by using <caption> - one indentation too much
  if (caption) code = code.replace(/\r\n\s{2}/g, '\r\n');
  // generate code block
  var res = caption + '```javascript\n' + code + '\n```';
  return res;
});

handlebars.registerHelper('desc', function(description) {
  // replace <code></code> block with gfm code block ``
  description = description.code();
  // remove <p></p> tags from description
  description = description.p();
  return description;
});

handlebars.registerHelper('list', function(param) {
  // split param names like config.key in array
  var names = param.name.split('.');
  // check for params like config.key
  var child = (names.length > 1) ? true : false;
  // indent if param is a child
  var indentation = child ? '  ' : '';
  // optional flag only for parent params
  var optional = (param.optional && !child) ? ' *optional* ' : '';
  var name = child ? names[1] : param.name;
  name = '`' + name + '`';
  var type = Array.isArray(param.type) ? param.type.join(', ') : param.type;
  type = '**' + type + '**';
  // remove <p></p> tags from description
  var desc = param.description.p();
  desc = ' - ' + desc;
  // default values
  var def = param.default ? ' - default `' + param.default + '`' : '';
  return indentation + '- ' + name + ' ' + type + optional + desc + def;
});

handlebars.registerHelper('return', function(returns) {
  if (!returns) return;
  var title = '#### Returns\n\n';
  var type = '- **' + returns.type + '**';
  var desc = returns.description ? ' - ' + returns.description.p().code() : '';
  var res = title + type + desc;
  return res;
});

// read handlebars template from source
var source = fs.readFileSync(__dirname + '/templates/template.handlebars', 'utf8');
var template = handlebars.compile(source);

// generate docs
module.exports = function() {
  // loop over functions
  docs.functions.forEach(function(fn) {
    // generate markdown file for each function
    var data = template(fn);
    // write markdown files
    fs.writeFileSync(__dirname + '/includes/' + fn.name + '.md', data);
  });
};


index.js
'use strict';

var url = require('url');
var base32 = require('thirty-two');
var totp = require('notp').totp;

/**
 * Prevent users who aren't logged-in from accessing routes.
 * Use `login.route` for redirection. Function also remembers the requested url
 * and user is redirected after successful login. If `rest` is enabled
 * you'll get a `401` response.
 *
 * @example
   `config.js`

   exports.login = {
     route: '/login'
   };
 *
 * @example
   `app.js`

   var config = require('./config.js');
   app.get('/private', utils.restrict(config), function(req, res) {
     res.send('only a logged in user can see this');
   })
 *
 * @param {Object} [config] - Configuration object
 * @param {String} [config.login.route='/login'] - Route that handles the login process
 */
exports.restrict = function(config) {

  config = config || {};
  var route = (config.login && config.login.route) || '/login';

  return function(req, res, next) {
    if (req.session.loggedIn) {return next(); }
    if (config.rest) {return res.send(401); }
    res.redirect(route + '?redirect=' + req.originalUrl);
  };

};

/**
 * Get type of database and database adapter name from connection information.
 *
 * @example
   `config.js (CouchDB)`

   exports.db = 'http://127.0.0.1:5984/';
 *
 * @example
   `config.js (all other DBs)`

   exports.db = {
     url: 'postgres://127.0.0.1:5432/',
     name: 'users',
     collection: 'my_user_table'
   }
 *
 * @example
   `app.js`

   var config = require('./config.js');
   var db = util.getDatabase(config);
   // {
   //   type: 'couchdb',
   //   adapter: 'lockit-couchdb-adapter'
   // }
 *
 * @param {Object} config - Configuration object
 * @param {String|Object} config.db - Database connection string / object
 *
 * @returns {Object} Object containing database `type` and `adapter`
 */
exports.getDatabase = function(config) {

  var uri = config.db.url || config.db;

  var urlObj = url.parse(uri);
  var prot = urlObj.protocol;

  var res = {};

  switch (prot) {
    case 'http:':
    case 'https:':
      res.type = 'couchdb';
      res.adapter = 'lockit-couchdb-adapter';
      break;
    case 'mongodb:':
      res.type = 'mongodb';
      res.adapter = 'lockit-mongodb-adapter';
      break;
    case 'postgres:':
      res.type = 'postgresql';
      res.adapter = 'lockit-sql-adapter';
      break;
    case 'mysql:':
      res.type = 'mysql';
      res.adapter = 'lockit-sql-adapter';
      break;
    case 'sqlite:':
      res.type = 'sqlite';
      res.adapter = 'lockit-sql-adapter';
      break;
  }

  return res;

};

/**
 * Generate link to QR code,  uses [Google Charts](https://developers.google.com/chart/infographics/docs/qr_codes).
 *
 * @example
   var config = {
     key: 'abcd1234',
     email: 'mirco.zeiss@gmail.com'
   };
   var link = util.qr(config);
   // https://chart.googleapis.com/chart?chs=200x200&cht=qr&chl=otpauth%3A%2F%2Ftotp%2FLockit%3Amirco.zeiss%40gmail.com%3Fsecret%3DMFRGGZBRGI2DI%3D%3D%3D%26issuer%3DLockit
 *
 * @param {Object} config - Configuration object
 * @param {String} config.key - Individual random key for user
 * @param {String} config.email - User email for Google Authenticator app
 * @param {String} [config.issuer='Lockit'] - Issuer for Google Authenticator
 *
 * @returns {String} URL for QR code
 */
exports.qr = function(config) {

  var key = config.key;
  var email = config.email;
  var issuer = config.issuer || 'Lockit';

  var label = issuer + ':' + email;
  var encoded = base32.encode(key);
  var uri = encodeURIComponent('otpauth://totp/' + label + '?secret=' + encoded + '&issuer=' + issuer);
  var api = 'https://chart.googleapis.com/chart?chs=200x200&cht=qr&chl=';
  return api + uri;

};

/**
 * Verify a two-factor authentication token, uses [time-based one-time password algorithm (totp)](http://en.wikipedia.org/wiki/Time-based_One-time_Password_Algorithm).
 * To be used with [Google Authenticator](https://support.google.com/accounts/answer/1066447?hl=en).
 *
 * @example
   var key = 'abcd1234';
   var token = '236709';
   var valid = util.verify(token, key);
   if (valid) {
     // continue here
   }
 *
 * @param {String} token - The two-factor token to verify
 * @param {String} key - The individual key for the user
 * @param {Object} [options] - Options object for
 *   [notp#totp.verify](https://github.com/guyht/notp#totpverifytoken-key-opt)
 * @param {String} [options.window=6] - Allowable margin for counter
 * @param {Number} [options.time=30] - Time step of counter in seconds
 *
 * @returns {Boolean} `true` if token is valid
 */
exports.verify = function(token, key, options) {

  options = options || {};

  var verified = totp.verify(token, key, options);
  return (verified && verified.delta === 0);

};

/**
 * Destroy the current session. Works with cookie sessions and session stores.
 *
 * @example
   util.destroy(req, function() {
     // user is now logged out
   });
 *
 * @param {Object} req - The default Express request object
 * @param {Function} done - Function executed when session is destroyed
 */
exports.destroy = function(req, done) {

  if (req.sessionStore) {return req.session.regenerate(done); }
  req.session = null;
  done();

};



// private helper function
function pipe(source, target) {
  var emit = source.emit;
  source.emit = function () {
    emit.apply(source, arguments);
    target.emit.apply(target, arguments);
    return source;
  };
}



/**
 * Pipe events from `source` to `target`. `source` can be a single event
 * emitter or an Array of event emitters.
 *
 * @example
   var util = require('util');
   var events = require('events');
   var utils = require('lockit-utils');

   var Child = function() {};
   util.inherits(Child, events.EventEmitter);

   var Mother = function() {};
   util.inherits(Mother, events.EventEmitter);

   var child = new Child();
   var mother = new Mother();

   utils.pipe(child, mother);

   mother.on('action', function(action) {
     console.log('look the child is ' + action);
   });

   child.emit('action', 'smiling');
 *
 * @param {Object|Array} source - Single event emitter or Array of event emitters
 * @param {Object} target - Single event emitter
 */
exports.pipe = function(source, target) {

  var isArray = Array.isArray(source);

  if (!isArray) {return pipe(source, target); }

  source.forEach(function(event) {
    pipe(event, target);
  });

};
